providers/azure.go

```diff
@@ -179,18 +180,26 @@ func (p *AzureProvider) Redeem(ctx context.Context, redirectURL, code string) (*
 
 // EnrichSession finds the email to enrich the session state
 func (p *AzureProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
-	if s.Email != "" {
-		return nil
-	}
+	if s.Email == "" {
 
-	email, err := p.getEmailFromProfileAPI(ctx, s.AccessToken)
-	if err != nil {
-		return fmt.Errorf("unable to get email address: %v", err)
+		email, err := p.getEmailFromProfileAPI(ctx, s.AccessToken)
+		if err != nil {
+			return fmt.Errorf("unable to get email address: %v", err)
+		}
+		if email == "" {
+			return errors.New("unable to get email address")
+		}
+		s.Email = email
 	}
-	if email == "" {
-		return errors.New("unable to get email address")
+
+	if s.User == "" {
+		user, err := p.getUserFromProfileAPI(ctx, s.AccessToken)
+		if err != nil {
+			return fmt.Errorf("unable to get user: %v", err)
+		}
+		// Not going to throw an error if this is empty
+		s.User = user
```

Comment:
Just looking at this, we are calling the profile URL twice here, is it possible we could make sure we only call the profile URL once and reuse the response?

Reply:
sorry can't pinpoint this one by looking at the code. line 185 and line 196 are calling two different functions. is that what you're referring to?

Reply:
Yeah, while one is get email and one is get user, they both under the hood call the userprofile API right? So is there a way we could make that a single request to the userprofile API rather than two? 

I think if we checked for both empty at the top of this function, then returned early if they aren't. Then we know we are going to need the request at least once so we can pull the info then pass the response into the other two functions so they can extract the info needed

Reply:
Ah yes I got your point. I can merge the two functions into something like `getEmailAndUserFromProfileAPI` and return both at the same time, and build the logic around `EnrichSession` to cater for empty values. Since these functions are not used anywhere else, don't see that making any issues. thoughts?